### ISSUE TITLE
Treatment of more than one s-expression in the repl

### ISSUE THREAD

**Issue by soegaard** [greghendershott/racket-mode#311]
The syntax highlighter only colors the first input s-expression.

Example:

```
foo.rkt﻿> (+ 1 2) (+ 3 4) (+ 5 6)       
3
foo.rkt﻿> 7
foo.rkt﻿> 11

```
Here only the `(+ 1 2)` in colored on the input line.
Also: Why are "foo.rkt" printed between 3, 7, and 11?

**Comment by greghendershott** [comment]
Two distinct things here:

1. The font-lock is due to mistakenly assuming only one sexpr will be present. (`comint-mode` does the bold text you see for the remaining sexprs; racket-mode undoes that, restoring the "nice" highlighting. But it's undoing only the first sexpr.) That's a simple fix.

2. As for the alternating input and prompts: You actually get that behavior with the vanilla `read-eval-print-loop`, because the default `current-read-interaction` is mostly just `read-syntax`. And e.g. `(with-input-from-string "(+ 1 2) (+ 3 4)" read-syntax)` is just `#<syntax::1 (+ 1 2)>`. In other words -- prompt/read/eval/print one sexpr at a time.

    You can see this trying your example with either `racket -e '(read-eval-print-loop)'` (no XREPL involved) or `TERM=dumb racket` (prevent XREPL from using `readline`).  It looks like XREPL tries to install for `current-prompt-read` the `read-cmdline-syntax` function from the `readline` collection. I'm guessing that does something more line-oriented that handles multiple sexprs. Likewise DrRacket seems to handle multi exprs.

    I'm not sure it's worth tinkering with this, especially if it could result in new bugs to find and fix?

**TL;DR**
1. is a bug that I'll fix.
2. IMHO is low-priority enhancement I might move to its own issue as a someday/maybe item.

**Comment by greghendershott** [comment]
Well I ended up addressing both points, after all. See last paragraph of the new doc string:

> RET (translated from \<return\>) runs the command
> racket-repl-eval-or-newline-and-indent (found in
> racket-repl-mode-map), which is an interactive Lisp closure in
> ‘racket-repl.el’.
>
> It is bound to RET.
>
> (racket-repl-eval-or-newline-and-indent)
>
> Evaluate or ‘newline-and-indent’.
>
> If you’ve supplied a complete s-expression, it is submitted to
> Racket to be evaluated. Otherwise, this does
> ‘newline-and-indent‘.
>
> - This is most useful when you are _not_ using something like
>   Paredit: Until you type the closing paren, you can use RET to
>   ‘newline-and-indent’.
>
> - If you _are_ using something like Paredit: You will probably
>   always have a complete s-expression. To enter one s-expression
>   over multiple lines, use something like C-j.
>
> **If you supply more than one s-expression, only the first is
> evaluated by Racket. The remainder move down to the next prompt,
> where you can evaluate the next one by pressing RET again.**

**Comment by greghendershott** [comment]
I'm re-opening this, because now I think the approach I took was wrong.

That approach has Emacs break them into multiple s-expressions, and `comint-send-input` them one by one to Racket input.  However:

- There were some bugs: #313 and #314. Although these are easy enough to fix, there's more...

- This breaks normal input to a program, in the REPL. e.g. You run a program that does `readline`, and type two words. They wrongly get split into two lines.

- Although I'm not sure if/when/how #305 will be fixed (I _think_ partly that depends on `datalog`??), this approach is probably going in the wrong direction -- for racket-repl-mode to be fussing with input assuming sexprs. It should limit itself to using the `drracket:submit-predicate`, if available.

---

What to do instead?

- As a fallback, I'll just effectively revert commit ab0abe6.

- Ideally I'd like to make this work like DrRacket and `xrepl`/`readline`/`racket/sandbox`. But I've been looking at them for longer than I'd care to admit, and still not quite grokking what they're doing. I'm not finding the needle in either of those hatystacks. They're not using the stock (prompt)`read-eval-print-loop`.  Instead is it:

    1. A `prompt`-`reads until #\newline`-`eval and print multiple expression`-`loop` ??  I tried to do this with stock `read-eval-print-loop`, by having a `current-prompt-read` handler

    2. A normal `read-eval-print-loop`, but the `current-prompt-read` handler optionally doesn't display the prompt if it sees another sexpr is already waiting in input ??

I tried to do 1 while sticking with the stock `read-eval-print-loop`, by changing the default `current-read-interaction` to something like this:

```racket

(define (sexpr-read-interaction source in)
  (parameterize ([read-accept-reader #t]
                 [read-accept-lang #f]
                 [current-input-port in])
    ;; Read at least one sexpr, skipping trailing whitespace after
    ;; each one, until we reach a #\newline.
    (define stxs
      (for/list ([stx (in-producer (read-syntax/skip-trailing-whitespace source))]
                 #:final (newline?))
        stx))
    #`(values #,@stxs)))

(current-read-interaction sexpr-read-interaction)

(define ((read-syntax/skip-trailing-whitespace source))
  (begin0 (read-syntax source)
    (skip-non-newline-whitespace)))

(define (newline?)
  (and (eq? #\newline (peek-char))
       (begin (read-char) #t)))

(define (skip-non-newline-whitespace)
  (when (non-newline-whitespace? (peek-char))
    (read-char)
    (skip-non-newline-whitespace)))

(define (non-newline-whitespace? c)
  (and (not (eq? c #\newline))
       (or (char-whitespace? c)
           (char-blank? c))))
```

Except how to wrap the values.

`begin` won't work; only the last value prints.

`values` (as above) seems to work -- except if you type something like `(values (values 1 2))` in the REPL.

Also this isn't wrapping each of them in `#%top-interaction`.  So, I guess I need to modify the stock `read-eval-print-loop`??

Or am I being dense and overlooking something simple/obvious?

**Comment by greghendershott** [comment]
Actually idea (ii) above seems promising. It works well with the `current-read-interaction` handler for sexprs.

```racket
(define/contract (make-prompt-read m)
  (-> (or/c #f mod?) (-> any))
  ;; A channel to which a thread puts interactions that it reads using
  ;; the current-read-interaction handler. This can be set by a lang
  ;; from its configure-runtime, so, this should be compatible with
  ;; any lang, even non-sexpr langs.
  (define ch (make-channel))
  (define (read-interaction/put-channel)
    (define in ((current-get-interaction-input-port)))
    (define v ((current-read-interaction) (object-name in) in))
    (channel-put ch v)
    (read-interaction/put-channel))
  (thread read-interaction/put-channel)
  ;; The prompt-read handler.
  (define (prompt-read)
    (match (sync/timeout 0 ch)
      ;; If no interaction ready, display prompt and wait.
      [#f
       (display-prompt (maybe-mod->prompt-string m))
       (channel-get ch)]
      ;; Another interaction ready, no need to prompt again.
      [v v]))
  prompt-read)
```
